geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/MessageDispatcher.java

```diff
@@ -432,34 +443,32 @@ protected void runDispatcher() {
           _messageQueue.remove();
           clientMessage = null;
         } catch (AuthenticationExpiredException expired) {
-          if (waitForReAuthenticationStartTime == -1) {
-            waitForReAuthenticationStartTime = System.currentTimeMillis();
-            // only send the message to clients who can handle the message
-            if (getProxy().getVersion().isNewerThanOrEqualTo(RE_AUTHENTICATION_START_VERSION)) {
-              EventID eventId = createEventId();
-              sendMessageDirectly(new ClientReAuthenticateMessage(eventId));
-            }
-            // We wait for all versions of clients to re-authenticate. For older clients we still
-            // wait, just in case client will perform some operations to
-            // trigger credential refresh on its own.
-            Thread.sleep(200);
-          } else {
+          waitForReAuthenticationStartTime = System.currentTimeMillis();
+          // only send the message to clients who can handle the message
+          if (getProxy().getVersion().isNewerThanOrEqualTo(RE_AUTHENTICATION_START_VERSION)) {
+            EventID eventId = createEventId();
+            sendMessageDirectly(new ClientReAuthenticateMessage(eventId));
+          }
+          // We wait for all versions of clients to re-authenticate. For older clients we still
+          // wait, just in case client will perform some operations to
+          // trigger credential refresh on its own.
+          synchronized (re_auth_lock) {
```

Comment:
Do you think it is possible that the response to the ClientReAuthenticateMessage is received before the MessageDispatcher starts waiting for notification?

Reply:
MessageDispatcher starts waiting immediately after sending the ClientReAuthenticateMessage. This seems very unlikely.

Reply:
You mean we should move the message sending into the synchronized block?

Reply:
I'm not sure. I was just thinking if a GC occurs after the message has been sent but before the wait occurs, it might miss it. I guess the GC would also cause the PutUserCredentials message from the client to not be processed either. So, I guess this is a non-issue.